app/services/cache_service.rb

```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+class CacheService
+  HOME_STATS_EXPIRY = 1.day.freeze
+  SHOPS_EXPIRY = 15.seconds.freeze
+
+  def self.cache(cache_key, options = {})
+    Rails.cache.fetch cache_key.to_s, options do
+      yield
+    end
+  end
+
+  # Yields a cached query, expired by the most recently updated record for a given class.
+  # E.g: if *any* Spree::Taxon record is updated, all keys based on Spree::Taxon will auto-expire.
+  def self.cached_data_by_class(cache_key, cached_class)
+    Rails.cache.fetch "#{cache_key}-#{cached_class}-#{latest_timestamp_by_class(cached_class)}" do
+      yield
+    end
+  end
+
+  # Gets the :updated_at value of the most recently updated record for a given class, and returns
+  # it as a timestamp, eg: `1583836069`.
+  def self.latest_timestamp_by_class(cached_class)
+    cached_class.maximum(:updated_at).to_i
+  end
+
+  def self.home_stats(statistic)
+    Rails.cache.fetch("home_stats_count_#{statistic}",
+                      expires_in: HOME_STATS_EXPIRY,
+                      race_condition_ttl: 10) do
+      yield
+    end
+  end
+
+  module FragmentCaching
```

Comment:
I wonder why is this mixed up with the previous code.
Looks like it would make a nice FragmentCacheService

Reply:
mmm doesn't this smell as merge conflict with your other PR? or am I looking at the wrong file?

Reply:
Yes, I split the original PR into 4, but actually they all need to touch things or add sections in the same files :see_no_evil:

Reply:
The merge conflicts should be very easy to resolve though.